content/blog/2019-10-3-tdd-gimmick/index.md

```diff
@@ -0,0 +1,196 @@
+
+Well yes it's a bit silly, but it highlights a central problem with unit testing. Namely:
+
+> How do you know that the test you've written is correct?
+
+Any alternative proposed to TDD **must** be able to solve this problem or else it cannot be said to be viable.
+
+The fact that the test is written first solves this problem. If the test is correct, it _must_ fail.
+
+In the example above you expect it to fail, so you'll revisit the test-code when you see it's passing even if no code is present. The rules push you to prioritize confidence in the test first before starting on the change you want to make.
+
+## Hmm... Well what about Logical Errors, TDD is not going to save me there
+
+TDD doesn't guarantee that your code is bug-free. It just guarantees that if a line is deleted that a test will fail. Therefore it's still your responsibility to write the correct tests. For example if the requirement was to multiply two numbers, but your tests are checking if the code adds the two together then your code is still wrong.
+
+In such a scenario, TDD helps in two ways:
+
+- it makes such an error easier to spot during code-review
+  - since your `expected` value will clearly be something different than the requirement, someone can flag it as an issue
+- once the error is spotted, it becomes easier to rewrite the bad test or to add tests for missing edge cases.
+
+Since refactoring is easier, this whole process happens much more quickly. And yes, while fixing one issue you're less likely to introduce a a regression even in this new feature.
+
+## This is all horribly complicated! I'm a pretty good programmer, I don't need all this
+
+I felt the same way when first starting out. Even now, when discussing some of the finer points of TDD it can feel a little overly-cautious and pessimistic. The thing is though, if there's more than 2 programmers working on a project it's just not possible to always be available for everyone.
+
+Imagine you've written a function, and someone wants to make a change to it. If they knew that the tests were high-quality, they could just figure it out themselves and be fairly confident that they weren't doing anything wrong. In the absence of a test-suite they would (hopefully) come and sit on your head and work things out with you.
+
+In the worst case, they're under a lot of pressure so they make the change that re-introduces some subtle bug that you'd already taken care of initially. The bug might only get discovered months later, at which point maybe the dev who introduced the problematic-change is on leave and the commit message just says "fixed issue" so now you're really screwed.
+
+At it's best, TDD is a way to communicate your decisions without needing to write extensive documentation that no one will ever read (or maintain when they make changes). Good-quality unit tests save such an incredible amount of time, and ultimately that's more time that you have to work in peace.
```

Comment:
I agree with this halfheartedly. Even if tests document the what, they do not document the why, which some times is what you need to decide if you need to fix the expectation of a test'

Reply:
Usually isn't the why just "because that was the requirement" ? So if the test description is written in terms of the product requirement then it should be easy to see what needs to be changed imo.

Reply:
Ref: https://github.com/jLouzado/patchcording.com/pull/11#issuecomment-541662742